### PR TITLE
cloud: Refactor managed machine pool

### LINKED ISSUE #757: [help wanted] refactor remaining services following established pattern
Following the pattern established in #716, we should refactor each service to follow the same pattern. Each service should have its own refactor PR to make reviewing easier. For example, network interfaces was already done in #742.

- [x] disks (@cpanato)

- [x] groups (@cpanato)

- [x] exp/managedclusters

- [ ] exp/agentpools

- [x]  public IPs (@CecileRobertMichon)

- [x] network interfaces (@CecileRobertMichon)

- [x] load balancers  (@CecileRobertMichon)

- [x] subnets (@cpanato)

- [x] virtual networks (@cpanato)

- [x] route tables (@nader-ziada)

- [x] inbound nat rules (@CecileRobertMichon)

- [x] role assignments (@CecileRobertMichon)

- [x]  resource skus (@alexeldeib)

- [x] securitygroups (@CecileRobertMichon)

- [x] exp/scalesets (@CecileRobertMichon)

- [x] virtualmachines (@CecileRobertMichon)

Let me know if you want to help with one of these in comments and I'll add your username next to it!

/help

### PR DESCRIPTION
**Pull request by nprokopic**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Follow up from #716, this PR applies the same pattern to managed machine pools.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #757

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
cloud: Refactor managed machine pool
```

### CODE REVIEW

**Comment by nprokopic** [comment]
cc @LochanRn @alexeldeib

File: azure/types.go [hunk 1]
```diff
@@ -358,8 +359,30 @@ type ManagedClusterSpec struct {
...
+	ResourceGroup string
+
+	// Cluster is the name of the AKS cluster.
+	Cluster string
+
+	// Version defines the desired Kubernetes version.
+	Version *string
+
+	// SKU defines Azure VM size for the agent pool VMs.
+	SKU string
```
Comment (alexeldeib) [inline azure/types.go:375]: nit: defines **the**

File: azure/types.go [hunk 2]
```diff
@@ -358,8 +359,30 @@ type ManagedClusterSpec struct {
...
+	// Replicas is the number of desired machines.
+	Replicas int32
+
+	// OSDiskSizeGB is OS disk size in GB for every machine in this agent pool.
 	OSDiskSizeGB int32
+
+	// VnetSubnetID is the Azure Resource ID for the subnet which should contain nodes.
+	VnetSubnetID string
+
+	// Mode represents mode of an agent pool. Possible values include: 'AgentPoolModeSystem', 'AgentPoolModeUser'.
```
Comment (alexeldeib) [inline azure/types.go:386]: these are the Azure values, not what we expose right? I guess doc'ing them here is fine, but they shouldn't be exposed really.

**Comment by alexeldeib** [comment]
/lgtm

very minor nits, looks good :) nice to finally wrap up all the refactoring!

File: azure/types.go (hunk 2 above)
Comment (nprokopic) [reply azure/types.go:386]: Oh right, thanks. I saw that we just use the same values like Azure SDK code, so I copied this from [code docs from Azure SDK](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2021-05-01/containerservice@v55.2.0+incompatible#ManagedClusterAgentPoolProfileProperties.Mode) (which are then not really correct strings, those are const names from the code :facepalm:).

File: azure/types.go (hunk 1 above)
Comment (nprokopic) [reply azure/types.go:375]: Fixed.

File: azure/types.go [hunk 3]
```diff
@@ -358,8 +359,30 @@ type ManagedClusterSpec struct {
...
+	// Version defines the desired Kubernetes version.
+	Version *string
+
+	// SKU defines Azure VM size for the agent pool VMs.
+	SKU string
+
+	// Replicas is the number of desired machines.
+	Replicas int32
+
+	// OSDiskSizeGB is OS disk size in GB for every machine in this agent pool.
```
[the remark this answers is not in the document]
Comment (nprokopic) [reply azure/types.go]: Fixed.

File: azure/types.go (hunk 2 above)
Comment (nprokopic) [reply azure/types.go]: Fixed.

Comment (nprokopic) [reply azure/types.go]: The values that are in the code doc now are the same as in `AzureManagedMachinePoolSpec.Mode`, since those are used here as well.

**Comment by alexeldeib** [comment]
hmm, maybe a potential flake for me to poke at further? I still had a few changes I was testing...

/test pull-cluster-api-provider-azure-e2e-exp

**Comment by alexeldeib** [comment]
force push diff: https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/37fabd41df7871ef7f8c85278bdfa26dacbd7cbb..da0975a4c8dfb4c68cb5f34972ef8f6fafbd6389

I'll assume the exp failure was a flake and debug it along with #1488 

/lgtm

**Comment by LochanRn** [comment]
The PR lgtm 👍

File: azure/interfaces.go [hunk 4]
```diff
@@ -31,14 +31,6 @@ type Reconciler interface {
 	Delete(ctx context.Context) error
 }
 
-// OldService is a generic interface for services that have not yet been refactored.
-// Once all services have been converted to use Service, this should be removed.
-// Example: virtualnetworks service would offer Reconcile/Delete methods.
-type OldService interface {
```
Comment (CecileRobertMichon) [inline azure/interfaces.go:37]: 🥳

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

File: azure/interfaces.go (hunk 4 above)
Comment (LochanRn) [reply azure/interfaces.go:37]: 💯

Comment (nprokopic) [reply azure/interfaces.go:37]: :tada:

**Pull request merged**